### PR TITLE
MM-42489_Revamp: Login screen webapp

### PR DESCRIPTION
**Pull request by julmondragon**
#### Summary
Apply new designs for the Login screen which should consider also cases when:
- Default login
- OAuth options are available
- Custom branding is configured
- MFA token

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-42489

#### Design
https://www.figma.com/file/7BM8S6BL597zBE83U6H8CK/MM-29159-Desktop-%26-Webapp-Login-Update

##### Login with username/email
Screen Shot 2022-05-09 at 12 41 16 PM

##### Login with username/email and external options
Screen Shot 2022-05-09 at 12 40 38 PM

##### Login with external options
Screen Shot 2022-05-09 at 12 45 21 PM

##### MFA enabled
mfa

###### Email verified
Screen Shot 2022-05-09 at 12 55 44 PM

###### Password updated
Screen Shot 2022-05-09 at 12 57 00 PM

###### Session expired
Screen Shot 2022-05-09 at 12 19 28 PM

###### Failed credentials
failure

###### No sign-in methods
Screen Shot 2022-05-09 at 1 00 13 PM

###### Default
May-05-2022 16-08-32

###### Custom branding
May-05-2022 15-52-53

#### Release Note
```release-note
Apply new designs for the Login screen:
- Default login
- OAuth options
- Custom branding
- MFA token
```

### CODE REVIEW

**Comment by julmondragon** [comment]
Currently working on update e2e.

**Comment by matthewbirtch** [comment]
@julmondragon I will take a closer look tomorrow likely but at first glance, I noticed the background visual is pretty low-res. I'm pretty positive this background already exists in the onboarding screens and we may be able to just use that. I think it may have even been implemented as an SVG.

I'll review everything else tomorrow in more detail though

**Comment by julmondragon** [comment]
> @julmondragon I will take a closer look tomorrow likely but at first glance, I noticed the background visual is pretty low-res. I'm pretty positive this background already exists in the onboarding screens and we may be able to just use that. I think it may have even been implemented as an SVG.
> 
> I'll review everything else tomorrow in more detail though

Login is reusing the same one onboarding process uses currently, it's in `jpg` format:(https://github.com/mattermost/mattermost-webapp/blob/master/images/admin-onboarding-background.jpg)
Just let us know if it needs to be updated. Thanks!

File: sass/utils/_mixins.scss [hunk 1]
```diff
@@ -126,23 +123,71 @@
 @mixin primary-button {
```
Comment (pvev) [inline sass/utils/_mixins.scss:123]: not completely sure, but I think this mixin is used in other places in the webapp, did you verified these changes won't affect the look of the button in other places where it is used?

File: components/widgets/inputs/password_input/password_input.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    onBlur?: FocusEventHandler<HTMLInputElement>;
+    onFocus?: FocusEventHandler<HTMLInputElement>;
+    hasError?: boolean;
+    error?: string;
+    createMode?: boolean;
+    disabled?: boolean;
+    inputSize?: SIZE;
+};
+
+const PasswordInput = React.forwardRef(({className, value, onChange, onBlur, onFocus, hasError, error, createMode, disabled, inputSize}: PasswordInputProps, ref?: React.Ref<HTMLInputElement>) => {
```
Comment (pvev) [inline components/widgets/inputs/password_input/password_input.tsx:25]: nit: place the variables in new lines so it is easier to read. Wondering why we don't have the maximum characters restriction per line in our linter.

File: components/widgets/inputs/input/input.stories.tsx [hunk 3]
```diff
@@ -242,6 +242,24 @@ storiesOf('Widgets/Inputs/Input', module).
                 );
             };
 
+            return (
+                <WrapperComponent/>
+            );
+        },
+    ).
+    add(
```
Comment (pvev) [inline components/widgets/inputs/input/input.stories.tsx:250]: nice!

File: components/widgets/icons/login_gitlab_icon.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import {useIntl} from 'react-intl';
+
+export default function LockCircleSolidIcon(props: React.HTMLAttributes<HTMLSpanElement>) {
+    const {formatMessage} = useIntl();
```
Comment (pvev) [inline components/widgets/icons/login_gitlab_icon.tsx:8]: nit: wondering if for all these icons instead of formating the aria-label message it sould make more sense to pass that value as a prop and avoid the extra import and also they would be more "customizable"

File: components/login/login.scss [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+            width: 610px;
+        }
+
+        .login-body-message .login-body-message-title {
+            padding-right: 130px;
+        }
+    }
+}
+
+@media screen and (max-width: 1199px) {
```
Comment (pvev) [inline components/login/login.scss:282]: are these breakpoints use in the media queries constant across the application? I mean, are those the ones we use and should keep using? if so, wouldn't it make sense to have it as css variables?

File: components/login/login.scss [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+        justify-content: center;
+
+        .login-body-no-login {
+            display: flex;
+            max-width: 540px;
+            flex-flow: column;
+            justify-content: center;
+
+            .login-body-no-login-title {
+                color: #1e325c;
```
Comment (pvev) [inline components/login/login.scss:25]: do we have a css variable for this? if not there are several occurrences of this hex in the PR too, maybe makes sense to create one

**Review by pvev — APPROVED** [review]
a couple of suggestions and questions, but in general really Nice work @julmondragon :shipit:

**Comment by JulienTant** [comment]
Question for PM/UX: Because in the figma design all the OpenID buttons have a white background, I wonder if we still want to give the option for admins to choose a background color for their OpenID buttons? If so, It would be nice to also offer them to change the text color as by default you are offered a blue background but the text in the button is dark-ish!

image

**Comment by julmondragon** [comment]
> Question for PM/UX: Because in the figma design all the OpenID buttons have a white background, I wonder if we still want to give the option for admins to choose a background color for their OpenID buttons? If so, It would be nice to also offer them to change the text color as by default you are offered a blue background but the text in the button is dark-ish!
> 
Relying on this on @matthewbirtch , but the scope of this work does not include deprecation or addition of features/options currently supported.

**Comment by matthewbirtch** [comment]
> Question for PM/UX: Because in the figma design all the OpenID buttons have a white background, I wonder if we still want to give the option for admins to choose a background color for their OpenID buttons? If so, It would be nice to also offer them to change the text color as by default you are offered a blue background but the text in the button is dark-ish!

Allowing these button colors to be customizable opens things up to a compromised design and could place emphasis on other buttons that weren't intended to have the emphasis.

Maybe we allow a custom color, but it only applies the colors to the text and button outline rather than the background?

**Comment by matthewbirtch** [comment]
Sorry for the duplicate message, not sure how that happened

File: components/alert_banner.tsx [hunk 7]
```diff
@@ -6,54 +6,53 @@ import classNames from 'classnames';
...
-                }
+const AlertBanner: React.FC<AlertBannerProps> = ({mode, title, message, className, variant = 'sys', onDismiss}: AlertBannerProps) => (
+    <div className={classNames('AlertBanner', mode, className, `AlertBanner--${variant}`)}>
+        <div className='AlertBanner__icon'>
+            <i
+                className={classNames({
+                    'icon-alert-outline': mode === 'danger' || mode === 'warning',
+                    'icon-check': mode === 'success',
+                    'icon-alert-circle-outline': mode === 'info',
+                })}
```
Comment (marianunez) [inline components/alert_banner.tsx:28]: Nice refactor 💯

File: components/login/login_controller/login_controller.test.jsx [hunk 8]
```diff
@@ -1,159 +0,0 @@
...
-
-import React from 'react';
-
-import {shallowWithIntl} from 'tests/helpers/intl-test-helper';
-
-import Constants from 'utils/constants';
-import LocalStorageStore from 'stores/local_storage_store';
-import LoginController from 'components/login/login_controller/login_controller';
-
-describe('components/login/LoginController', () => {
```
Comment (marianunez) [inline components/login/login_controller/login_controller.test.jsx:12]: Did we move all this test cases for the loginController somewhere ?

File: components/preparing_workspace/preparing_workspace.scss [hunk 9]
```diff
@@ -53,12 +53,14 @@
 
     .tertiary-button {
         @include tertiary-button;
+        @include button-medium;
```
Comment (marianunez) [inline components/preparing_workspace/preparing_workspace.scss:56]: Are these changes intended here?

**Review by marianunez — COMMENTED** [review]
Awesome work @julmondragon! That was a lot of cases to cover 🚀

File: components/alert_banner.scss [hunk 10]
```diff
@@ -1,46 +1,49 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
 .AlertBanner {
     display: flex;
     overflow: hidden;
-    padding: 16px;
+    align-items: flex-start;
+    padding: 14px;
```
[the remark this answers is not in the document]
Comment (julmondragon) [reply components/alert_banner.scss:8]: Yes, ensured that current use cases with different modes available continued working as expected.  Adding storybook could help.

File: components/login/login.tsx [hunk 11]
```diff
@@ -0,0 +1,806 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+/* eslint-disable max-lines */
+import React, {useState, useEffect, useRef, useCallback} from 'react';
```
[the remark this answers is not in the document]
Comment (julmondragon) [reply components/login/login.tsx:5]: Correct, basically this is the same login_controller rewritten in TS, all workflows and logic are mainly unchanged.
Improvements were contained to the render elements and more clear validations for each use case.

**Comment by JulienTant** [comment]
> Relying on this on @matthewbirtch , but the scope of this work does not include deprecation or addition of features/options currently supported.

👍🏻 yep I was just opening the discussion!

File: components/preparing_workspace/preparing_workspace.scss (hunk 9 above)
Comment (julmondragon) [reply components/preparing_workspace/preparing_workspace.scss:56]: Yes, it is needed to specify styling according to size as well. As mentioned in other comments, for the next iteration of this (Signup) plan to review the option to have a button wrapper component, so that type and sizes could be set via props.

File: e2e/cypress/tests/integration/system_console/site_configuration/link_customization_e20_1_spec.js [hunk 12]
```diff
@@ -10,6 +10,7 @@
 // Stage: @prod
 // Group: @system_console @enterprise @e20_only @not_cloud
 
+import * as TIMEOUTS from '../../fixtures/timeouts';
```
Comment (Phrynobatrachus) [inline e2e/cypress/tests/integration/system_console/site_configuration/link_customization_e20_1_spec.js:13]:
```suggestion
import * as TIMEOUTS from '../../../fixtures/timeouts';
```

File: e2e/cypress/tests/integration/system_console/site_configuration/link_customization_cloud_spec.js [hunk 13]
```diff
@@ -10,6 +10,7 @@
 // Stage: @prod
 // Group: @system_console @enterprise @cloud_only
 
+import * as TIMEOUTS from '../../fixtures/timeouts';
```
Comment (Phrynobatrachus) [inline e2e/cypress/tests/integration/system_console/site_configuration/link_customization_cloud_spec.js:13]:
```suggestion
import * as TIMEOUTS from '../../../fixtures/timeouts';
```

**Review by Phrynobatrachus — CHANGES_REQUESTED** [review]

File: e2e/cypress/tests/integration/system_console/site_configuration/link_customization_e20_1_spec.js (hunk 12 above)
Comment (julmondragon) [reply e2e/cypress/tests/integration/system_console/site_configuration/link_customization_e20_1_spec.js]: Updated!

File: e2e/cypress/tests/integration/system_console/site_configuration/link_customization_cloud_spec.js (hunk 13 above)
Comment (julmondragon) [reply e2e/cypress/tests/integration/system_console/site_configuration/link_customization_cloud_spec.js]: Updated!

File: components/login/login_controller/login_controller.test.jsx (hunk 8 above)
Comment (marianunez) [reply components/login/login_controller/login_controller.test.jsx:12]: Perfect, I had missed it with some many files

**Comment by julmondragon** [comment]
> Great work on this @julmondragon. I didn't find a whole lot. A few comments beyond the minor comments I made on code:
> ...

1. Looks like this is not possible in the cloud spinwick. We need to ensure this is tested in self managed as part of QA work. finally, there are e2e test that were updated covering this.
2. Per discussion with the team this will be worked separately. Main ticket is https://mattermost.atlassian.net/browse/MM-23853, will create specific ticket if needed.
3. Updated styles to handle larger images.
5. Updated handling of Custom Description per our review.

cc @matthewbirtch

File: components/preparing_workspace/preparing_workspace.scss (hunk 9 above)
Comment (julmondragon) [reply components/preparing_workspace/preparing_workspace.scss:56]: Yes, the styles for mixin `tertiary-button` were updated, so need the size ones added here.

Comment (marianunez) [reply components/preparing_workspace/preparing_workspace.scss:56]: Gotcha, thanks for clarifying!

**Review by Phrynobatrachus — APPROVED** [review]
https://automation-dashboard.vercel.app/cycles/68279927-588d-4e26-ab37-b4cf1d587740

Latest run seems to be solid, thanks!

**Review by marianunez — APPROVED** [review]
Thanks @julmondragon - great work 🚀

**Review by JulienTant — APPROVED** [review]
LGTM @julmondragon good job!!

**Review by stevemudie — APPROVED** [review]
I'll test this post merge.

**Review by matthewbirtch — APPROVED** [review]
